onnxruntime/core/providers/webgpu/webgpu_context.cc

```diff
@@ -0,0 +1,677 @@
+                                                  })) == Status::OK());
+      auto mapped_data = static_cast<const uint64_t*>(query_read_buffer.GetConstMappedRange());
+
+      for (size_t i = 0; i < pending_kernels.size(); i++) {
+        const PendingKernelInfo& pending_kernel_info = pending_kernels[i];
+        const auto& inputs = pending_kernel_info.inputs;
+        const auto& outputs = pending_kernel_info.outputs;
+
+        SS(shapes, 128);
+        for (size_t s = 0; s < inputs.size(); s++) {
+          const auto& input = inputs[s];
+          shapes << "inputs[" << s << "] = " << input.override_shape.ToString() << " ";
+        }
+        for (size_t s = 0; s < outputs.size(); s++) {
+          const auto& output = outputs[s];
+          shapes << "outputs[" << s << "] = " << output.override_shape.ToString() << " ";
+        }
+
+        if (gpu_timestamp_offset_ == 0) {
+          gpu_timestamp_offset_ = mapped_data[i * 2];
+          // TODO: apply CPU-GPU time offset so that timestamps are aligned
+        }
+        uint64_t start_time = mapped_data[i * 2] - gpu_timestamp_offset_;
+        uint64_t end_time = mapped_data[i * 2 + 1] - gpu_timestamp_offset_;
+
+        const std::unordered_map<std::string, std::string>& event_args = {
+            {"shapes", SS_GET(shapes)},
+            {"cache_key", pending_kernel_info.cache_key},
+        };
+
+        profiling::EventRecord event(profiling::API_EVENT,
+                                     -1,
+                                     -1,
+                                     pending_kernel_info.name,
+                                     static_cast<int64_t>(std::round(start_time / 1000.0)),
+                                     static_cast<int64_t>(std::round((end_time - start_time) / 1000.0)),
+                                     event_args);
+        events.emplace_back(std::move(event));
+      }
+
+      query_read_buffer.Unmap();
+      query_read_buffer.Destroy();
+    }
+
+    pending_queries_.clear();
+  }
+
+  is_profiling_ = false;
+}
+
+void WebGpuContext::EndProfiling(TimePoint /* tp */, profiling::Events& events, profiling::Events& cached_events) {
+  // This function is called when no active inference is ongoing.
+  ORT_ENFORCE(!is_profiling_, "Profiling is ongoing in an inference run.");
+
+  if (query_type_ != TimestampQueryType::None) {
+    // No pending kernels or queries should be present at this point. They should have been collected in CollectProfilingData.
+    ORT_ENFORCE(pending_kernels_.empty() && pending_queries_.empty(), "Pending kernels or queries are not empty.");
+
+    events.insert(events.end(),
+                  std::make_move_iterator(cached_events.begin()),
+                  std::make_move_iterator(cached_events.end()));
+
+    cached_events.clear();
+  } else {
+    LOGS_DEFAULT(WARNING) << "TimestampQuery is not supported in this device.";
```

Comment:
>LOGS_DEFAULT(WARNING) [](http://example.com/codeflow?start=4&length=21)

Using the session logger would be better

Reply:
I try to make a simple change to this, but it turns out to be not that easy. It requires to make multiple changes to pass the reference of the session logger through call chains so that I can use it here.

working on it.

Reply:
whether TimestampQuery is supported or not depends on session. Is it still considered better to use a session logger here?